packages/services/src/rewards.service.ts

```diff
@@ -51,15 +45,46 @@ export class RewardsService {
   }
 
   /**
-   * Calculate rewards for a given input
+   * Calculate rewards and allocate them for a competition
+   *
+   * This method combines the reward calculation and allocation process:
+   * 1. Calculates rewards using the internal calculation logic
+   * 2. Stores the calculated rewards in the database
+   * 3. Allocates the rewards by building a Merkle tree and publishing to blockchain
+   *
+   * @param competitionId The competition ID (UUID) to calculate and allocate rewards for
+   * @param startTimestamp The timestamp from which rewards can be claimed
+   * @throws Error if reward calculation fails or no rewards exist to allocate
+   */
+  public async calculateAndAllocate(
+    competitionId: string,
+    startTimestamp: number,
+  ): Promise<void> {
+    const prizePool =
+      await this.competitionRepository.getCompetitionPrizePools(competitionId);
+    if (!prizePool) {
+      throw new Error(`No prize pool found for competition ${competitionId}`);
+    }
+
+    await this.calculateRewards(
```

Comment:
Here we calculate the rewards using the `rewards` package and store them in the database table `rewards`